components/daikin_s21/s21.cpp

```diff
@@ -192,224 +147,476 @@ std::string str_repr(uint8_t *bytes, size_t len) {
+      uint8_t byte;
+      rx_uart->read_byte(&byte);
+      last_event_time = millis();
+      result = handle_rx(byte);
+    }
+    break;
   }
-  this->write_frame(code);
 
-  uint8_t byte;
-  if (!this->rx_uart->read_byte(&byte)) {
-    ESP_LOGW(TAG, "Timeout waiting for %s response", c.c_str());
-    return false;
+  return result;
+}
+
+DaikinSerial::Result DaikinSerial::send_frame(const char *cmd, const std::array<char, S21_PAYLOAD_SIZE> *payload) {
+  if (comm_state != CommState::Idle) {
+    return Result::Busy;
   }
-  if (byte == NAK) {
-    ESP_LOGD(TAG, "NAK from S21 for %s query", c.c_str());
-    return false;
+
+  size_t cmd_len = strlen(cmd);
+  if (cmd_len > S21_MAX_COMMAND_SIZE) {
+    ESP_LOGW(TAG, "Tx: Command '%s' too large", cmd);
+    return Result::Error;
   }
-  if (byte != ACK) {
-    ESP_LOGW(TAG, "No ACK from S21 for %s query", c.c_str());
-    return false;
+
+  if (debug) {
+    if (payload == nullptr) {
+      ESP_LOGD(TAG, "Tx: %s", cmd);
+    } else {
+      ESP_LOGD(TAG, "Tx: %s %s %s", cmd,
+               str_repr(reinterpret_cast<const uint8_t*>(payload->data()), payload->size()).c_str(),
+               hex_repr(reinterpret_cast<const uint8_t*>(payload->data()), payload->size()).c_str());
+    }
   }
 
-  std::vector<uint8_t> frame;
-  if (!this->read_frame(frame)) {
-    ESP_LOGW(TAG, "Failed reading %s response frame", c.c_str());
-    return false;
+  // prepare for response
+  response.clear();
+  flush_input();
+
+  // transmit
+  tx_uart->write_byte(STX);
+  tx_uart->write_array(reinterpret_cast<const uint8_t *>(cmd), cmd_len);
+  uint8_t checksum = std::accumulate(cmd, cmd + cmd_len, 0U);
+  if (payload) {
+    tx_uart->write_array(reinterpret_cast<const uint8_t *>(payload->data()), payload->size());
+    checksum = std::accumulate(payload->data(), payload->data() + payload->size(), checksum);
+  }
+  if (checksum == STX) {
+    checksum = ENQ;  // mid-message STX characters are escaped
   }
+  tx_uart->write_byte(checksum);
+  tx_uart->write_byte(ETX);
 
-  this->tx_uart->write_byte(ACK);
+  // wait for result
+  last_event_time = millis();
+  comm_state = (payload != nullptr) ? CommState::CommandAck : CommState::QueryAck;
 
-  std::vector<uint8_t> rcode;
-  std::vector<uint8_t> payload;
-  for (size_t i = 0; i < frame.size(); i++) {
-    if (i < code.size()) {
-      rcode.push_back(frame[i]);
-    } else {
-      payload.push_back(frame[i]);
+  return Result::Ack;
+}
+
+void DaikinSerial::flush_input() {  // would be a nice ESPHome API improvement
+  uint8_t byte;
+  while (rx_uart->available()) {
+    rx_uart->read_byte(&byte);
+  }
+}
+
+void DaikinS21::refine_queries() {
+  // exclude F9 if unit supports the individual sensors with better resolution
+  if (support_Ra && support_RH)
+  {
+    const auto f9 = std::find_if(std::begin(queries), std::end(queries), [](const char * query){ return strcmp(query, "F9") == 0; });
```

Comment:
This is my lazy proof of concept for cleaning up the query pool. My units don't support F9, it's already removed by the Nak handling. Could avoid the search every scan if we set a flag for each refinement we make.